rllib/examples/rlhf/ppo_ft/rlhf_env.py

```diff
@@ -0,0 +1,220 @@
+        # This is the SFT model, used for computing the kl-divergence penalties.
+        self.sft_model = transformers.GPT2LMHeadModel.from_pretrained(
+            config["sft_model_path"]
+        )
+
+        # This is the reward model, it's either a LLM or a callable function 
+        # that takes in a batch of generated text tokens + response mask and returns a 
+        # batch of rewards.
+        # TODO(Kourosh): Later allow users to pass in their own reward model factory.
+        self.reward_model = ShortestAnswerRM(max_length=self.max_generation_length)
+
+        # Prompt dataset
+        # We need to load the prompt dataset and randomly sample a prompt from 
+        # it upon reset.
+        # TODO (Kourosh): Assuming we only support HF datasets for now:
+        self.prompt_dataset = datasets.load_dataset(
+            config["prompt_dataset_path"], 
+            split=config["prompt_dataset_split"]
+        )
+        self.dsize = len(self.prompt_dataset)
+
+
+        vocab_size = self.tokenizer.vocab_size
+        model_max_length = self.tokenizer.model_max_length
+
+        self.action_space = sp.Dict({
+            "sequence": Repeated(sp.Discrete(vocab_size), max_len=model_max_length),
+            "attention_mask": Repeated(sp.Discrete(2), max_len=model_max_length),
+            "response_mask": Repeated(sp.Discrete(2), max_len=model_max_length),
+            "logits": Repeated(
+                sp.Box(0, 1, shape=(vocab_size,)), max_len=model_max_length
+            ),
+        })
+
+        self.observation_space = sp.Dict({
+            "input_ids": Repeated(sp.Discrete(vocab_size), max_len=model_max_length),
+            "attention_mask": Repeated(sp.Discrete(2), max_len=model_max_length),
+        })
+
+
+    def reset(self, *, seed=None, options=None):
+        if seed:
+            np.random.seed(seed)
+        index = np.random.randint(self.dsize)
+        prompt = self.prompt_dataset[index]["prompt"]
+        prompt_tokens = self.tokenizer(prompt, return_tensors="np")
+        # remove the batch dimension since we can only do one sentence generation at a 
+        # time.
+        prompt_tokens = tree.map_structure(lambda x: x[0], prompt_tokens)
+
+        return prompt_tokens, {}
+
+    def step(self, action):
+        # action is a dict with the following keys:
+        # sequence: [b, NP+NR], prompt + generated response from the actor LLM
+        # response_mask: [b, NP+NR], [0] * NP + [1] * NR To separate response from prompt
+        # attention_mask: [b, NP+NR]
+        # probs: [b, NR, vocab_size], the probabilities of each token in the vocab for each token in the input
+
+        sequence = action["sequence"]
+        response_mask = action["response_mask"]
+        attention_mask = action["attention_mask"]
+        logits = action["logits"]
+
+        n_response_tokens = response_mask.sum()
+
+        r_align = self.reward_model(sequence, attention_mask, response_mask)
+        r_align = r_align.item()
+
```

Comment:
Maybe you can add a configurable discount to the reward, as done [here](https://github.com/nebuly-ai/nebullvm/blob/013084714b7f26ca41afc4662df65432689abcfe/apps/accelerate/chatllama/chatllama/rlhf/trainer.py#L962)

```python
gamma = self.config.trainer.gamma_discounted
discounted_rewards = torch.zeros_like(old_values)
for i in range(discounted_rewards.shape[1]):
    for j in range(i, discounted_rewards.shape[1]):
        discounted_rewards[:, i] += (
            gamma ** (j - i) * rewards[:, j]
        )
```